services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,531 @@
+
+        return getForCertificateAuthority(subjectDn, span);
+    }
+
+    /**
+     * Finds an existing tenant by the subject DN of its configured certificate authority.
+     *
+     * @param subjectDn the subject DN to find the tenant with.
+     * @param span The active OpenTracing span for this operation. It is not to be closed in this method!
+     *            An implementation should log (error) events on this span and it may set tags
+     *            and use this span as the parent for any spans created in this method.
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>200 OK</em> if a tenant with the given certificate authority is registered.
+     *         The <em>payload</em> will contain the tenant's configuration information.</li>
+     *         <li><em>400 Bad Request</em> if the certificate authority is missing.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier.</li>
+     *         </ul>
+     */
+    private Future<TenantResult<JsonObject>> getForCertificateAuthority(final X500Principal subjectDn,
+            final Span span) {
+        if (subjectDn == null) {
+            TracingHelper.logError(span, "missing subject DN");
+            return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            final Future<Map.Entry<String, Versioned<Tenant>>> tenantsFound = getByCa(subjectDn);
+            return tenantsFound.compose(successTenantRead -> {
+                if (successTenantRead == null) {
+                    TracingHelper.logError(span, "no tenant found for subject DN");
+                    return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                } else {
+                    return Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils.convertTenant(successTenantRead.getKey(),
+                                    successTenantRead.getValue().getValue(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                }
+            });
+        }
+    }
+
+    /**
+     * Fetches tenant by subject DN.
+     *
+     * @param subjectDn the subject DN.
+     * @return A Future Map consisting of pair of a tenant id and a versioned tenant.
+     */
+    private Future<Map.Entry<String, Versioned<Tenant>>> getByCa(final X500Principal subjectDn) {
+
+        if (subjectDn == null) {
+            return Future.failedFuture("missing subject DN");
+        } else {
+            final JsonObject findTenantQuery = new MongoDbDocumentBuilder()
+                    .withCa("tenant", subjectDn.getName())
+                    .document();
+
+            final Promise<JsonObject> tenantsFound = Promise.promise();
```

Comment:
As I noticed in several places, please use a suitable name for variables of type `Promise`. Here I think `findTenantPromise`suits more than the `tenantsFound`. Also please take a look and make changes in other places too.

Reply:
done